cpp/src/endpoint.cpp

```diff
@@ -157,6 +165,8 @@ std::shared_ptr<Request> Endpoint::close(const bool enablePythonFuture,
                                          EndpointCloseCallbackUserFunction callbackFunction,
                                          EndpointCloseCallbackUserData callbackData)
 {
+  if (_callbackData->closing->exchange(true) || _handle == nullptr) return nullptr;
+
```

Comment:
> OK, so this means it is safe to call `close` multiple times. Did you figure out the circumstances under which `Endpoint::close` gets called from multiple places on the same endpoint? I assume that it can't _only_ be called from the dtor...

Yes, the user may call it as well multiple times so we need to ensure we don't attempt closing it after it's been closed or while/after `errorCallback()` runs.

Reply:
> Minor suggestion, use [`std::atomic_flag`](https://en.cppreference.com/w/cpp/atomic/atomic_flag) rather than `std::atomic<bool>` and use `_callbackData->closing.test_and_set()` instead of `exchange(true)`?

Good idea, done in https://github.com/rapidsai/ucxx/pull/184/commits/fbe93ea1cb35f292d8a64f8111cd993493c1c3f5 .

EDIT: Reverted this change, see comment replying https://github.com/rapidsai/ucxx/pull/184/files#r1547624363 .